geode-redis/src/main/java/org/apache/geode/redis/internal/RedisLockService.java

```diff
@@ -70,29 +71,38 @@ public AutoCloseableLock lock(ByteArrayWrapper key) throws InterruptedException
       throw new IllegalArgumentException("key cannot be null");
     }
 
+    KeyHashIdentifier lockKey = new KeyHashIdentifier(key.toBytes());
     Lock lock = new ReentrantLock();
-    Lock oldLock = map.putIfAbsent(key, lock);
+    Lock oldLock = weakReferencesTolocks.putIfAbsent(lockKey, lock);
+
     if (oldLock != null) {
       lock = oldLock;
-      // we need to get a reference to the actual key object so that the backing WeakHashMap does
-      // not clean it up.
-      for (ByteArrayWrapper keyInSet : map.keySet()) {
-        if (keyInSet.equals(key)) {
-          key = keyInSet;
-          break;
-        }
-      }
+
+      // we need to get a reference to the actual key object
+      // so that the backing WeakHashMap does not clean it up
+      // when garbage collection happens.
+      lockKey = getReferenceToLockKey(lockKey);
```

Comment:
I don't understand this logic. If `weakReferencesTolocks.putIfAbsent(lockKey, lock)` returns a value, then you already have a reference to the lockKey. OR am I mistaken about this?

Reply:
The essence here is that you need to get a reference to the actual key in the map (if you don't it might get GC'd). Consider the first pass through all of this and I'm trying to lock key "ABC" and succeed. I have a reference to the key and all is good. If another thread comes to look for key "ABC" it will have a different 'key' object. So if we get the lock out of the map but don't also get the actual key then that lock might get GC'd if we don't retain a hard reference to its key.

Reply:
What I was trying to say was... in the java doc for `putIfAbsent`... 

> If the specified key is not already associated with a value (or is mapped
     * to {@code null}) associates it with the given value and returns
     * {@code null}, else returns the current value.

Which makes the whole loop of iterating over the keySet just extra work.

Reply:
Correction... it return the VALUE not the KEY.... ok.... LGTM